metapool/metapool.py

```diff
@@ -1599,76 +1601,227 @@ def add_controls(plate_df, blanks_dir, katharoseq_dir):
+            # Katharoseq
+            plate_df['Sample'] = np.where(
+                (plate_df['Sample'].isna()) &
+                (plate_df == 'positive_control').any(axis=1),
+                "kathseq." + plate_df['number_of_cells'].astype(str) + "." +
+                plate_df['Row'] + plate_df['Col'].astype(str),
+                plate_df['Sample'])
+
+        elif katharoseq_dir is None:
+
+            # Merge plate_df with controls table
+            plate_df = pd.merge(plate_df, blanks, on='TubeCode', how='left')
 
         # Assign sample_names ('Sample') to controls
+        # BLANKS
         plate_df['Sample'] = np.where(
             (plate_df['Sample'].isna()) &
             (plate_df == 'negative_control').any(axis=1),
             "BLANK" + plate_df['Project Plate'].str.split('_').str.get(-1) +
             "." + plate_df['Row'] + plate_df['Col'].astype(str),
             plate_df['Sample'])
 
-        plate_df['Sample'] = np.where(
-            (plate_df['Sample'].isna()) &
-            (plate_df == 'positive_control').any(axis=1),
-            "kathseq." + plate_df['number_of_cells'].astype(str) + "." +
-            plate_df['Row'] + plate_df['Col'].astype(str), plate_df['Sample'])
-
         # Assign BLANK column
         plate_df['Blank'] = np.where(plate_df['Sample'].str.contains('BLANK'),
                                      True, False)
 
         print('Controls added')
 
         return plate_df
+
+
+def validate_plate_df(plate_df, metadata, sample_accession_df, blanks_dir,
+                      katharoseq_dir=None):
+    """"Function checks that all the samples names recorded in the plate_df
+    have metadata associated with them. It also checks that all the matrix
+    tubes in the plate_df are indeed located in the sample accesion file or
+    controls lists.Checks for duplicate sample names and makes sure the
+    Katharoseq Dilution curve has the full dilution series (8-point serial
+    dilution)
+    ----------
+    Args:
+    plate_df:
+        pandas DataFrame object
+    metadata:
+        pandas DataFrame of qiita metadata associated with study
+    blanks dir: dir
+        "*.tsv" files of tube IDs assigned to blank tubes
+    katharoseq_dir: dir (default:None)
+        "*_tube_ids.tsv", contains tube ids of katharoseq samples
+        "*_cells_counts.tsv", contains cell counts of katharoseq samples
+
+    Returns
+    -------
+    Returns 0 (zero)
+    """
+
+    # This checks that all the samples names recored in the plate_df have
+    # metadata associated with them
+    str_ = 'positive_control|negative_control'
+    control_samples = \
+        set(plate_df.loc[(plate_df['description'].str.contains(str_)) |
+                         (plate_df['description'].isna()), 'Sample'])
+
+    print(f"There are {len(control_samples)} control samples in this plate")
+    samples_in_metadata = []
+    if 'tube_id' in metadata.columns:
+        mask = plate_df['Sample'].isin(metadata['sample_name'].str[6:]) |\
```

Comment:
The '6' is kind of ambiguous; it might be good to add a comment showing an example so it's easier to understand why a substring starting at 6 is reliable. We need to know if there's a possibility that the string could change and it might need a value other than 6 as well. In that case, a method like str.substr() or re.search() might be better choices.

Reply:
I think you are right that the str[6:] slicing might not be perfect. It is intended to slice out the Qiita_ID prefix for sample_names in the metadata. The slicing gets rid of the 5 digit Qiita_ID and the period. Do you have another suggestion to achieve the same result. Alternatively, I added a comment to explain why the string slicing is being used. If you think that is sufficient, I'll leave the string slicing as is.